XrdCeph changes from STFC/RAL

Jo-stfc opened the pull request:
rebased PR of https://github.com/xrootd/xrootd/pull/2628

Jo-stfc committed:
[XrdCeph] add useratpool in server logs in case of failure

Jo-stfc committed:
[XrdCeph] clear checksum on write completion

Commit:
[XrdCeph] Add initial support for calculating Adler32 on-the-fly

alex-rg committed:
[XrdCeph] feat: allow buffering layer to work only for writes
It was noticed that for reads there may be some inefficiencies when
xcache is involved. For writes, however, the buffering layer improved
pefromance significantly.

So, it make sense to introduce the write-only mode.

Commit:
[XrdCeph] Change config options. The default is to not calculate the streamed checksum. Settings for streamed checksum allow calculation only, calc+log, and calc+log+store as extended attribute. Record streamed and readback checksums in a central checksums.log file.

Commit:
[XrdCeph] Open checksum record file once, in XrdCephPosix.cc . Do not log uninitialised checksum value when client reads file. Move recording of readback checksum value into XrdCephPosix instead of Set() method in XrdCephXAttr.cc. More checks for writable checksum record file

Commit:
[XrdCeph]Use strdup to ensure that concatenation of C String .c_str() works correctly

Commit:
[XrdCeph] Remove duplicate definition of ts_rfc3339 in XrdCephOss

Commit:
[XrdCeph] Move to use Use XRootD's own checksum calculator class rather than using zlib directly

Move from using zlib adler32() function to XrdCks/XrdCksCalcadler32. The
superclass, XrdCksCalc, provides a uniform interface to all supported
checksum algorithms, and is the preferred method in XRootD.
Remove an unneccessary guard condition in the ceph_posix_*write()
methods.

Commit:
[XrdCeph] Add a delete call on XrdCksCalcalder32 object obtained by new

The XrdCksCalc object in the CephFileRef structure needs a matching
delete in ceph_posix_close()

Jo-stfc committed:
[XrdCeph] linting improvements and disable streamed checksums for aio as it would be too risky

Jo-stfc committed:
[XrdCeph] add README for XrdCeph

Jo-stfc committed:
[XrdCeph] fix CI warning by removing <=0 char comparison

amadio reviewed (changes requested):
In its current form, I don't think we can merge this. Please have a look at the comments I left, and please check your code for memory leaks, I found a few just while reading, but I'm sure that are more. We can try to get this into XRootD 6.1 if there's no time to do it for 6.0. I think it's important to address the issues first. Thanks!

Jo-stfc committed:
[XrdCeph] memory leak fixes in streamed checksum code following PR requests
use strings instead of char * copies for streamed checksum code
rebase config load to only check log file path if streamed checksum config is enabled
remove duplicate extern declaration

Jo-stfc commented on src/XrdCeph/XrdCephXAttr.cc:
```diff
@@ -28,6 +28,9 @@
 #include "XrdCeph/XrdCephXAttr.hh"
 
+#include "XrdCks/XrdCksData.hh"
+#include <string.h>
```
resolved in latest commit

Jo-stfc commented on src/XrdCeph/XrdCephPosix.cc:
```diff
@@ -54,7 +54,50 @@
+     asciiVal[j++] = hex2ascii(loNibble(hexbyte));
+
+  }
+  return strdup(asciiVal);
```
refactored to use std::string, along with the following memory leaks

Jo-stfc commented on src/XrdCeph/XrdCephPosix.cc:
```diff
@@ -777,6 +852,30 @@ int ceph_posix_close(int fd) {
+  	  logwrapper((char*)"ceph_close: fd: %d, Adler32 streamed checksum = %s", fd, adler32Cks);
+
+      if (g_logStreamedAdler32) {
+	    const char *path = strdup((fr->pool + ":" + fr->name).c_str());
```
fixed in latest commit

amadio commented:
@Jo-stfc Do you still plan to work on this? I'd suggest we close this one, and you then reopen one rebased on XRootD 6.0 when it passes the current CI testing and the hackish code has been removed. Cheers,

Jo-stfc commented:
ok I'll reopen a new one after rebase

amadio commented:
Alright, thanks! I will close this one then. Please try to structure the commits in one logical change per commit to make reviewing easier if you can, and maybe create a separate pull request for just the addition of documentation, as that has much less strict requirements to get merged.
